packages/x-data-grid/src/hooks/features/rows/useGridRows.ts

```diff
@@ -215,10 +215,57 @@ export const useGridRows = (
 
   const updateNestedRows = React.useCallback<GridRowProPrivateApi['updateNestedRows']>(
     (updates, groupKeys) => {
-      const nonPinnedRowsUpdates = computeRowsUpdates(apiRef, updates, props.getRowId);
+      const insertedNodes = new Set<GridRowId>();
+      const nonPinnedRowsUpdates = computeRowsUpdates(
+        apiRef,
+        updates,
+        props.getRowId,
+        insertedNodes,
+      );
+
+      const tree = gridRowTreeSelector(apiRef);
+      const removedNodes = new Set<GridRowId>();
+
+      if (groupKeys && groupKeys.length > 0) {
+        const rootNode = tree[GRID_ROOT_GROUP_ID];
+        let parentNode = rootNode;
+        for (let i = 0; i < groupKeys.length; i += 1) {
+          const childrenFromPath = (parentNode as GridGroupNode).childrenFromPath;
+          const nodeId = childrenFromPath[Object.keys(childrenFromPath)[0]]?.[groupKeys[i]];
+          if (nodeId) {
+            parentNode = tree[nodeId];
+          }
+        }
+
+        const traverse = (node: GridGroupNode) => {
```

Comment:
Why is this whole traverse necessary?

Reply:
It is to check if the children of the row being fetched are outdated and need to be removed before adding new ones.

Regarding the part where parentId is derived from groupKeys: we could update the function signature to directly accept groupId. However, since the current approach isn’t expensive, I chose not to modify the signature. That said, I’m open to discussion if you think it’s better off directly accepting a param.